Future Work Page Draft

samimak37 opened the pull request:
Adds the Future Work Page to the site, along with the write-up associated with it.

In addition to the writeup, there is also a visualization of the nearest neighbor results in the form of a census map with colorings based on a given tract's similarity to Deanwood. The K-Means Graph is still a work in progress, but will be up soon.

Co-Authored with @JusticeV452.

samimak37 committed:
Created Future Work Page

samimak37 committed:
Renamed `future_work` dir to `future`

samimak37 committed:
Basic 1940 Census Map

Implements a basic 1940 Census Map using data provided by social explorer. Census tracts can be colored individually

samimak37 committed:
Implemented Similarity Census Map

Created a map of the 1940s census that displays tracts based on how "similar" they are to Deanwood

samimak37 committed:
Create Popups for Census Tracts

Created popups for the individual census tracts in `CensusTractMap`

JusticeV452 committed:
Merge in preliminary scatter plot

JusticeV452 committed:
Merge branch 'future-work-mvp' of https://github.com/dhmit/self-sustaining-cities into future-work-mvp

samimak37 committed:
Added Introduction and Datasets

samimak37 committed:
Merge branch 'future-work-mvp' of https://github.com/dhmit/self-sustaining-cities into future-work-mvp

samimak37 committed:
Merged scatter and map code

samimak37 committed:
Removed duplicated code

samimak37 committed:
Updated Scatter Plot Formatting

samimak37 committed:
Added wording for KMeans and Nearest Neighbor

JusticeV452 committed:
Added second radio buttons and update button names

JusticeV452 committed:
Merge branch 'future-work-mvp' of https://github.com/dhmit/self-sustaining-cities into future-work-mvp

samimak37 committed:
Added adversarial model writeup

samimak37 committed:
Added default tract focuses

Created default "fly to" locations for the similarity map

JusticeV452 committed:
Update checkbox orientations and fixed checkbox checking

JusticeV452 committed:
Merge branch 'future-work-mvp' of https://github.com/dhmit/self-sustaining-cities into future-work-mvp

samimak37 committed:
Removed trailing comma

samimak37 committed:
Merge branch 'main' into future-work-mvp

ryaanahmed commented on frontend/components/maps/CensusTractMap.js:
```diff
@@ -0,0 +1,179 @@
+                    />
+                    {this.props.census_tracts && this.props.deanwood_similarities &&
+                        this.props.census_tracts.map(tract => {
+                            return (
```
@anastasia mentioned a performance problem on this page, and it stems from here. Right now, on each render, `styleTracts` and `onEachFeature` is getting called for each of ~7500 tracts, and that's causing the page to be quite laggy.

You could try a few things...

Something like:
```
    shouldComponentUpdate(nextProps, _nextState, _nextContext) {
        return (nextProps !== this.props);
    }
```
will get rid of a bunch of unnecessary renders. e.g., switching the city selector right now causes an unnecessary rerender.

That doesn't get rid of an initial relatively slow loop through all of the tracts. Some thoughts there:
- it looks like there's a bunch of work happening in the frontend here that could be precomputed on the backend
- can you defer adding a GeoJSON to the MapContainer until it's actually in the viewport?

samimak37 replied:
I tinkered with it a bit, and I think I got the performance back to something reasonable. I couldn't find a way to prevent adding the GeoJSON objects until they were in the viewport (at least, not without querying the backend), but I was able to make a lot of the computation done in advance and just stored in the json object that gets sent to the front end. I also ended up cutting the number of tracts that the frontend deals with about in half (basically I made it such that the GeoJSON feature isn't even sent to the frontend if it isn't within a certain threshold). Hopefully all of this ends up helping, but please let me know! It's a lot more fluid on my computer, but I'm not sure how it'll hold up on others.

samimak37 committed:
Improved performance of census map

Improved the performance of the census tract map by moving calculations to be stored in the json objects instead of calculated at run time

samimak37 committed:
Merge branch 'main' into future-work-mvp

samimak37 committed:
Filled in results for `FutureWorkOverview`

ryaanahmed commented:
Awesome, yes -- I think this works. The problem wasn't so much the number of objects as the fact that there were lots of render calls recomputing a bunch of stuff that could've been static. Precomputing the stuff and preventing the extra renders improves things a lot on my machine, too.

samimak37 committed:
Added Census data and classifications to backend

Added the json files and endpoints for accessing kmeans data for future work on the backend

samimak37 committed:
Added missing encoding to `get_1940_tract_data`

JusticeV452 committed:
Merge branch 'main' into future-work-mvp

anastasia committed:
Merge branch 'main' into future-work-mvp

anastasia commented:
Thank you for figuring out the performance issues! Great work! Merging
